velox/exec/tests/utils/QueryAssertions.cpp

```diff
@@ -549,52 +549,291 @@ std::string toString(const MaterializedRow& row) {
     message << "\t";
-    printRow(extraActualRows[i], message);
+    printRow(extraRows[i], message);
     message << std::endl;
   }
   message << std::endl;
 
-  auto missingRowsToPrint = std::min((size_t)10, missingActualRows.size());
+  auto missingRowsToPrint = std::min((size_t)10, missingRows.size());
   message << missingRowsToPrint << " of missing rows:" << std::endl;
   for (int32_t i = 0; i < missingRowsToPrint; i++) {
     message << "\t";
-    printRow(missingActualRows[i], message);
+    printRow(missingRows[i], message);
     message << std::endl;
   }
   message << std::endl;
   return message.str();
 }
 
+std::string MaterializedRowEpsilonComparator::getUserFriendlyDiff() const {
+  VELOX_CHECK(
+      notEqual_, "This method must be called after compare() returned false.");
+
+  std::vector<MaterializedRow> extraRows;
+  std::vector<MaterializedRow> missingRows;
+  int32_t actualIndex = 0;
+  int32_t expectedIndex = 0;
+
+  while (expectedIndex < expectedSorted_.size() &&
+         actualIndex < actualSorted_.size()) {
+    const auto& expectedRow = expectedSorted_[expectedIndex];
+    const auto& actualRow = actualSorted_[actualIndex];
+    if (equalWithEpsilon(expectedRow, actualRow)) {
+      ++expectedIndex;
+      ++actualIndex;
+    } else if (lessThanWithEpsilon(expectedRow, actualRow)) {
+      missingRows.push_back(expectedRow);
+      ++expectedIndex;
+    } else {
+      extraRows.push_back(actualRow);
+      ++actualIndex;
+    }
+  }
+  for (; actualIndex < actualSorted_.size(); ++actualIndex) {
+    extraRows.push_back(actualSorted_[actualIndex]);
+  }
+  for (; expectedIndex < expectedSorted_.size(); ++expectedIndex) {
+    missingRows.push_back(expectedSorted_[expectedIndex]);
+  }
+
+  return makeErrorMessage(
+      missingRows, extraRows, expectedSorted_.size(), actualSorted_.size());
+}
+
+bool MaterializedRowEpsilonComparator::equalKeys(
+    const MaterializedRow& left,
+    const MaterializedRow& right) const {
+  for (auto i = 0; i < columns_.size() - numFloatingPointColumns_; ++i) {
+    auto column = columns_[i];
+    if (left[column] != right[column]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool MaterializedRowEpsilonComparator::hasUniqueKeys(
+    const std::vector<MaterializedRow>& sortedRows) const {
+  for (auto i = 1; i < sortedRows.size(); ++i) {
+    if (equalKeys(sortedRows[i], sortedRows[i - 1])) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool MaterializedRowEpsilonComparator::sortByUniqueKey(
+    const MaterializedRowMultiset& expected,
+    const MaterializedRowMultiset& actual) {
+  std::copy(
+      expected.begin(), expected.end(), std::back_inserter(expectedSorted_));
+  std::copy(actual.begin(), actual.end(), std::back_inserter(actualSorted_));
+  auto lessThan = [&](const MaterializedRow& lhs, const MaterializedRow& rhs) {
+    return customLessThan(
+        lhs,
+        rhs,
+        [](const variant& lhs, const variant& rhs) { return lhs == rhs; },
```

Comment:
== and < operators for non-floating point values are the same as equalsWithEpsilon and lessThanWithEpsilon. Hence, we can always to xxxWithEpsilon. If we do that we won't need customLessThan that takes equals and lessThan lambdas.

Reply:
The comparison here applies to floating-point columns too. We need strict comparison to avoid undefined behavior of sorting (as required by std library).